src/java.base/share/classes/jdk/internal/foreign/abi/x64/sysv/TypeClass.java

```diff
@@ -174,6 +176,7 @@ private static List<ArgumentClassImpl> classifyStructType(GroupLayout type) {
     }
 
     static TypeClass classifyLayout(MemoryLayout type) {
+        SharedUtils.checkHasNaturalAlignment(type);
```

Comment:
Isn't this check too deep in the call chain? E.g. isn't this some validation we could do immediately in the linker, by looking at the function descriptor being passed? Does it need to live in platform dependent code?

Reply:
Good point, it should be moved up. I went through an iteration where I only checked ValueLayouts, for which it made more sense to put here.